base/util/PKICertImport.bash

```diff
@@ -124,6 +277,27 @@ function PKICertImport() {
                 # Always required; certificate usage flags.
                 CERT_USAGE="$1"
                 shift
+            elif [ "x$arg" == "x--pkcs12-password" ] || [ "x$arg" == "x-w" ]; then
+                # If specified, password file for the .p12 file.
+                PKCS12_PASSWORD="$1"
+                shift
+            elif [ "x$arg" == "x--unsafe-keep-keys" ]; then
+                # If specified, keep keys imported from .p12 even when
+                # verification failed.
+                __e "Warning: --unsafe-keep-keys has been specified."
+                __e "This option allows you to keep keys imported from a .p12 when verification failed."
+                PKCS12_REMOVE_KEYS="false"
+            elif [ "x$arg" == "x--unsafe-trust-then-verify" ]; then
```

Comment:
what's the difference between "--chain-trust" and "--unsafe-trust-then-verify"?

Reply:
`--chain-trust <chain-trust-value>` specifies the trust to be used on intermediate certificates (i.e., all those which aren't the leaf). By default, setting this trust occurs **after** verification of the certificate and helps to ensure things chain to a trusted root (otherwise, a self-signed root certificate included in the `.p12` file would be marked trusted, its signature would be valid, and then it could be used to validate the leaf certificate). When `--unsafe-trust-then-verify` (which doesn't take any trust flags) is specified, we apply this trust *prior* to verification of certificate signatures, which allows us to do precisely that: import a root CA that isn't otherwise present in the NSS DB trust store.

That is, if we're given a `.p12` with three certs:

1. A root cert, `root.crt`
2. An intermediate chained to the root, `sub.crt`
3. A server certificate, `server.crt`

Then, our default mode is equivalent to:

1. `certutil -A -i root.crt -t ,,` -- no trust
2. `certutil -V -n root -u <chain-usage-value>` -- if fails, exit; it'll fail when the root certificate isn't already present, as it isn't marked as trusted in the previous step (so if it is actually added as a new certificate because it doesn't exist, it would be untrusted).
3. `certutil -M -n root -t <chain-trust-value>` -- the value passed after `--chain-trust`
4. `certutil -A -i sub.crt -t ,,` -- no trust
5. `certutil -V -n sub -u <chain-usage-value>` -- if fails, exit
6. `certutil -M -n sub -t <chain-trust-value>` -- the value passed after `--chain-trust`
7. `certutil -A -i server.crt -t ,,` -- no trust
8. `certutil -V -n server -u <usage-value>` -- if fails, exit
9. `certutil -M -n sub -t <trust-value>` -- the value passed after `--trust` (or `-t`).

Our `--unsafe-trust-then-verify` mode is equivalent to:
1. `certutil -A -i root.crt -t <chain-trust-value>` -- the value passed after `--chain-trust`
2. `certutil -V -n root -u <chain-usage-value>` -- if fails, exit; this'll pass if the trust flags are correct and the usage is valid and the key has a valid signature. Since we set trust flags in the previous step based on what the user specified, the user can mark all of these certificates as trusted, allowing them to import new root certificates.
3. `certutil -A -i sub.crt -t <chain-trust-value>` -- the value passed after `--chain-trust`
4. `certutil -V -n sub -u <chain-usage-value>` -- if fails, exit
5. `certutil -A -i server.crt -t <trust-value>` -- the value passed after `--trust` (or `-t`)
6. `certutil -V -n server -u <usage-value>` -- if fails, exit

Hopefully that makes sense... :)

Reply:
Yes that makes sense.  Although I think you have a typo on step 9 of --chain-trust default mode, where you said
certutil -M -n sub ...
I think you meant
certutil -M -n server ...
;-)